---
title: "Analysis of <i>Ascochyta rabiei</i> Population and Pathogenicity using DArT-Sequencing"
author: "Ido Bar"
date: "25 September 2019"
always_allow_html: yes
output: 
    # md_document:
#      css: "style/style.css"
      # toc: true
      # toc_depth: 3
#      highlight: pygments
#      number_sections: false
    html_document:
      css: "style/style.css"
      toc: true
      toc_float: true
      toc_depth: 3
      highlight: pygments
      number_sections: false
      code_folding: hide
#      keep_md: true
bibliography: style/RAD-seq.bib
csl: style/springer-basic-improved-author-date-with-italic-et-al-period.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(list(echo = TRUE, eval=FALSE, message=FALSE))
# load custom functions from github
devtools::source_gist("7f63547158ecdbacf31b54a58af0d1cc", filename = "util.R")
# options(width = 180)
cran_packages <- c("tidyverse", "knitr", "pander","kableExtra", "captioner", "DT", "htmltab",
                   "paletteer")
pacman::p_load(char=cran_packages, repos="https://cran.rstudio.com/")
# Connect to Zotero to access references
# biblio <- ReadBib("data/Fungal_genomes.bib") # , "bibtex", "RefManageR"
# Font Format
custom_font="consolas"
fontFmt = function(x,font="consolas"){
  #outputFormat = knitr::opts_knit$get("rmarkdown.pandoc.to")
  #if (outputFormat == 'html')
  formatted_text <- sprintf("<font face='%s'>%s</font>",font,x)
  return(formatted_text)
  #else
  #  x
}
```



```{r captions, include=FALSE, eval=TRUE}
figs <- captioner(prefix="Figure")
figs(name="GC_cont", "GC content per sample in WGS reads.")
appdx <- captioner(prefix="Appendix")
tbls <- captioner(prefix="Table")
tbls(name="sum_table","Summary of Australian isolates used for DArT sequencing across the years and regions (state).")
appdx(name="samples","Ascochyta rabiei isolates used for DArT sequencing.")
tbls(name="mapping_rates", "Mapping rates of the WGS reads to the <i>Botrytis cinerea</i> B05.10 reference genome.")
tbls(name="mapping_sum", "Mapping statistics.")
```

# Experimental Design
DNA was extracted from 279 isolates of Australian _Ascochyta rabiei_ and 2 Spanish _A. rabiei_ isolates and sent for DArT-Sequencing at Diversity Array Technologies (DArT, Canberra) on a single Illumina NovaSeq flowcell, producing 150 bp short paired-end reads (run name CAGRF20074).  
A summary of the Australian isolates used for DArT sequencing across the years and regions (state) is provided below (`r tbls(name="sum_table",display="cite")`).  
Additional details of the sequenced isolates is provided in `r appdx(name="samples",display="cite")`.

`r tbls(name="sum_table")`
```{r summary_table, eval=TRUE} 
isolate_table <- read_csv("data/DArT_metadata.csv") %>% rename(Isolate=id)
isolate_summary <- isolate_table %>% filter(State!="SPAIN") %>% count(Year, State) %>% 
  pivot_wider(names_from = State, values_from = n) %>% mutate_all(~replace_na(., 0)) %>% 
  mutate(Total=rowSums(.[-1]), Year=as.character(Year)) %>% 
  rbind(., c(NA, as.double(summarise_if(., is_double, ~sum(.))))) %>% mutate(Year=replace_na(Year, "Total"))

# datatable(as.data.frame(samples_table), caption=tbls("samples")) 

kable(isolate_summary) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"),
                full_width = F, position = "left") %>% 
  # column_spec(3, italic=TRUE) %>% 
  column_spec(7, bold = T) %>% 
  row_spec(7, bold = TRUE)
```

# Aims
* Identify patterns unique to _A. rabiei_ isolates collected within growing region (State), Year, Host cultivar
* Find association between genomic variants/haplotypes and pathogenicity level (Pathotype)
* Assess population structure (try unsupervised clustering and also by State/Year)

## Analysis Pipeline
### General overview:
1. Read in Data (after cross-validation of sample names in input files)
2. Data filtration:
    a. Filter by repeatability (`gl.filter.repavg()` in dartR) - a meassurement of quality per locus
    b. Filter by monomorphic loci (`gl.filter.monomorphs()`) as they do not provide information for population structure and simply slow the analysis
    c. Filter by amount of missing data (`gl.filter.callrate(method="loc")`) per locus
    d. Filter to remove all but one of multiple snps in the same fragment (`gl.filter.secondaries()`) - not needed for _fineradstructure_ analysis  
    e. Filter individuals by amount of missing data (`gl.filter.callrate(method="ind")`)
3. Perform population structure analysis using Poppr, StrataG, StockR or external tool such as ADMIXTURE/STRUCTURE (and their deriviatives)
  a. Calculate genetic distance  
  b. Create Tree for dendrogram  
  c. DAPC/PCA analysis

### Plots:
1. Dendrogram with all samples (including Spain) - use `pheatmap` to label Year and State of collection
2. DAPC/PCA of Australian isolates, colour by State/Year/Host/Pathogenicity


### Methods
DArT-Seq data processing, mapping and variant calling were performed on the _Griffith University Gowonda HPC Cluster_ (using Torque PBS scheduler) and the [EcoCloud](https://app.ecocloud.org.au/) cloud-based analysis platform.  
Detailed methods, including code for running each of the analysis steps are provided in the associated [A_rabiei_DArT GitHub repository](https://github.com/IdoBar/Botrytis_cinerea_probe_alignment).

#### Environment setup
Install needed software in a `conda` environment on a Linux machine (GU's Gowonda HPC cluster or [EcoCloud](https://app.ecocloud.org.au/) in this case).
```{bash setup_tools}
# install RAD tools (need to fix internet connection to gowonda2 - use patched netcheck in ~/bin)
~/bin/netcheck
# add and set priorities for conda channels
conda config --add channels conda-forge
conda config --append channels bioconda # not neccesary
conda config --append channels anaconda
# create an environment to process the data
conda create -n rad-seq snpsift snpeff blast samtools bcftools stacks ipyrad tassel plink admixtools admixture structure eigensoft
conda create -n fs -c genomedk finestructure libgomp fineradstructure 
# fix library dependency
ln -s /home/jovyan/workspace/.conda/finestruct/lib/libcblas.so /home/jovyan/workspace/.conda/finestruct/lib/libcblas.so.3
conda create -n faststruct faststructure
conda install -n rad-seq -c biobuilds igv
# Clean extra space
conda clean -y --all

```

#### Data pre-processing
Variants are provided by DArT-Seq in `.csv` files. The SNPs are in either 1-row or 2-row files while the _silico-dart_ file contains presence/absence information (whether the RE site was found or not in the sample).  
These files can be imported into `r fontFmt("R")` (@ihakaLanguageDataAnalysis1996, @R_2017) as `genlight` objects (@jombartAdegenet31New2011) using the `dartR` package (@gruberDartRPackageFacilitate2018) 

#### Population structure
Using `r fontFmt("fineradstructure")` (@malinskyRADpainterFineRADstructurePopulation2017, @lawsonInferencePopulationStructure2012, see [manual](http://cichlid.gurdon.cam.ac.uk/fineRADstructure.html) and [GitHub repo](https://github.com/millanek/fineRADstructure)) to infer the co-ancestry matrix and use it to cluster samples by populations.  

```{bash finerad}
# download scripts from the github repo https://github.com/millanek/fineRADstructure
# Calculate LD between loci and sort accordingly
Rscript scripts/sampleLD.R A_rabiei_DArT_fineradstructure.tsv A_rabiei_DArT_fineradstructure_LD.tsv
# Calculate the co-ancestry matrix
RADpainter paint A_rabiei_DArT_fineradstructure_LD.tsv
# Assign individuals to populations
finestructure -x 100000 -y 100000 -z 1000 A_rabiei_DArT_fineradstructure_LD_chunks.out A_rabiei_DArT_fineradstructure_LD_chunks.mcmc.xml
# build the tree (see issue https://github.com/millanek/fineRADstructure/issues/3)
fs fs -m T -x 10000 A_rabiei_DArT_fineradstructure_LD_chunks.out A_rabiei_DArT_fineradstructure_LD_chunks.mcmc.xml A_rabiei_DArT_fineradstructure_LD_chunks.mcmcTree.xml
```

## Appendices
```{r samples_table, eval=TRUE} 
datatable(as.data.frame(isolate_table), caption=appdx(name="samples")) 
# kable(isolate_table) %>%
#   kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"),
#                 full_width = TRUE, position = "left") %>% 
#   # column_spec(3, italic=TRUE) %>% 
#   column_spec(1, bold = T)
```

### Appendix 2. Useful resources



## General information
This document was last updated at `r Sys.time()` using R Markdown (built with `r R.version.string`). Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. It is especially powerful at authoring documents and reports which include code and can execute code and use the results in the output. For more details on using R Markdown see <http://rmarkdown.rstudio.com> and [Rmarkdown cheatsheet](https://www.rstudio.com/wp-content/uploads/2016/03/rmarkdown-cheatsheet-2.0.pdf).

***
## Bibliography



